python/mlx/nn/utils.py

```diff
@@ -37,34 +37,46 @@ def wrapped_value_grad_fn(*args, **kwargs):
     return wrapped_value_grad_fn
 
 
-def checkpoint(module: Module, fn: Callable = None):
```

Comment:
There is a lot of python magic going on here but with a second look I think it is fine. Ie not that hidden/hard to reason about.

The only thing is that it is not entirely clear to me what happens when someone keeps around both modules and starts messing with them. Would we expect changes in one to be reflected to the other?

Reply:
> The only thing is that it is not entirely clear to me what happens when someone keeps around both modules and starts messing with them. Would we expect changes in one to be reflected to the other?

I think the behavior to get across is that it makes a *new* module and because of that you should not see changes reflected in the old one. Now I'm only 90% certain that's the case :), but if it's not it is a bug and I should add a test / fix it.